src/tls.rs

```diff
@@ -327,20 +351,53 @@ impl<T: Transport + Unpin> Transport for TlsStream<T> {
     fn remote_addr(&self) -> Option<SocketAddr> {
         self.io.inner.remote_addr()
     }
+
+    fn upgrade_to_https(&self) -> bool {
+        !self.is_tls
+    }
 }
 
 pub(crate) struct TlsAcceptor {
     config: Arc<ServerConfig>,
     incoming: AddrIncoming,
+    upgrade: bool,
+    socket: Option<AddrStream>,
 }
 
 impl TlsAcceptor {
-    pub(crate) fn new(config: ServerConfig, incoming: AddrIncoming) -> TlsAcceptor {
+    pub(crate) fn new(config: ServerConfig, incoming: AddrIncoming, upgrade: bool) -> TlsAcceptor {
         TlsAcceptor {
             config: Arc::new(config),
             incoming,
+            upgrade,
+            socket: None,
         }
     }
+
+    /// stream directly to `TlsStream`
+    fn poll_stream(
+        &mut self,
+        sock: AddrStream,
+        is_tls: bool,
+    ) -> Poll<Option<Result<TlsStream<AddrStream>, io::Error>>> {
+        let session = ServerSession::new(&self.config.clone());
+        // let tls = Arc::new($this.config);
+        Poll::Ready(Some(Ok(TlsStream::new(sock, session, is_tls))))
+    }
+}
+
+/// check for valid HTTPS before moving to `TlsStream`
+macro_rules! poll_check {
+    ($self:expr, $cx:expr, $sock:expr) => {{
+        let mut first_byte = [0];
+        match ready!($sock.poll_peek($cx, &mut first_byte)) {
```

Comment:
This will keep peeking at the first byte until `Ready`, until then it will keep `Pending`.
It doesn't need a timeout or anything, I'm assuming this is handled somewhere else?

Reply:
This is also now handled in `tokio-rustls`.